crypto/zk/circuit/circuit.go

```diff
@@ -0,0 +1,278 @@
+	if err := circuit.LoadRemote(ctx); err != nil {
+		return nil, err
+	}
+
+	// checks hashes of current files
+	if correct, err := circuit.VerifiedCircuitArtifacts(); err != nil {
+		return nil, err
+	} else if !correct {
+		return nil, fmt.Errorf("download artifacts does not match with the expected ones")
+	}
+
+	return circuit, nil
+}
+
+// LoadLocal tries to read the content of current circuit artifacts from its
+// local path (provingKey, verificationKey and wasm). If any of the read
+// operatios fails, returns an error.
+func (circuit *ZkCircuit) LoadLocal() error {
+	var err error
+
+	log.Debugw("loading circuit locally...", map[string]interface{}{
+		"localDir": circuit.Config.LocalDir})
+
+	// compose files localpath
+	provingKeyLocalPath := filepath.Join(circuit.Config.LocalDir,
+		circuit.Config.CircuitPath, circuit.Config.ProvingKeyFilename)
+	verificationKeyLocalPath := filepath.Join(circuit.Config.LocalDir,
+		circuit.Config.CircuitPath, circuit.Config.VerificationKeyFilename)
+	wasmLocalPath := filepath.Join(circuit.Config.LocalDir,
+		circuit.Config.CircuitPath, circuit.Config.WasmFilename)
+
+	// read file contents into circuit parameters
+	circuit.ProvingKey, err = os.ReadFile(provingKeyLocalPath)
+	if err != nil {
+		return fmt.Errorf("error reading provingKey locally: %w", err)
+	}
+
+	circuit.VerificationKey, err = os.ReadFile(verificationKeyLocalPath)
+	if err != nil {
+		return fmt.Errorf("error reading verificationKey locally: %w", err)
+	}
+
+	circuit.Wasm, err = os.ReadFile(wasmLocalPath)
+	if err != nil {
+		return fmt.Errorf("error reading wasm circuit locally: %w", err)
+	}
+
+	return nil
+}
+
+// LoadRemote downloads the content of the current circuit artifacts from its
+// remote location. If any of the downloads fails, returns an error.
+func (circuit *ZkCircuit) LoadRemote(ctx context.Context) error {
+	log.Debugw("circuit not downloaded yet, downloading...", map[string]interface{}{
+		"localDir": circuit.Config.LocalDir})
+
+	baseUri, err := url.Parse(circuit.Config.URI)
+	if err != nil {
+		return err
+	}
+
+	remotePath := fmt.Sprintf("%s/%s", baseUri.String(), circuit.Config.CircuitPath)
+	localPath := filepath.Join(circuit.Config.LocalDir, circuit.Config.CircuitPath)
+	if err := os.MkdirAll(localPath, os.ModePerm); err != nil {
+		return err
+	}
+
+	// Compose provingKey remote and local locations
+	provingKeyUri := fmt.Sprintf("%s/%s", remotePath, circuit.Config.ProvingKeyFilename)
+	provingKeyLocalPath := filepath.Join(localPath, circuit.Config.ProvingKeyFilename)
+	// Compose verificationKey remote and local locations
+	verificationKeyUri := fmt.Sprintf("%s/%s", remotePath, circuit.Config.VerificationKeyFilename)
+	verificationKeyLocalPath := filepath.Join(localPath, circuit.Config.VerificationKeyFilename)
+	// Compose wasm remote and local locations
+	wasmUri := fmt.Sprintf("%s/%s", remotePath, circuit.Config.WasmFilename)
+	wasmLocalPath := filepath.Join(localPath, circuit.Config.WasmFilename)
```

Comment:
this vars init could certainly be refactored to be reused in LoadLocal

and even built into some struct that lets you then do some kind of `for _, key := range keys`
```
	// Download and store locally key
	key.dest, err = downloadFile(ctx, key.uri)
	if err != nil {
		return fmt.Errorf("error downloading %s: %w", key.name, err)
	}
	if err := storeFile(key.dest, key.localPath); err != nil {
		return fmt.Errorf("error storing %s: %w", key.name, err)
	}
	```

Reply:
instead of the triplicated code below

Reply:
Agree! I tried to implement changes to improve the code in this way. Please, check it @altergui and let me know your opinion.